apps/emqx/src/emqx_passwd.erl

```diff
@@ -139,11 +131,41 @@ compare_secure([X | RestX], [Y | RestY], Result) ->
 compare_secure([], [], Result) ->
     Result == 0.
 
+-define(MAX_DERIVED_KEY_LENGTH, (1 bsl 32 - 1)).
 pbkdf2(MacFun, Password, Salt, Iterations, undefined) ->
-    pbkdf2:pbkdf2(MacFun, Password, Salt, Iterations);
+    crypto:pbkdf2_hmac(MacFun, Password, Salt, Iterations, dk_length(MacFun));
+pbkdf2(_MacFun, _Password, _Salt, _Iterations, DKLength) when DKLength > ?MAX_DERIVED_KEY_LENGTH ->
+    %% backword compatible with the old implementation in pbkdf2
+    error(dklength_too_long);
 pbkdf2(MacFun, Password, Salt, Iterations, DKLength) ->
-    pbkdf2:pbkdf2(MacFun, Password, Salt, Iterations, DKLength).
+    crypto:pbkdf2_hmac(MacFun, Password, Salt, Iterations, DKLength).
 
 hex(X) when is_binary(X) ->
     %% TODO: change to binary:encode_hex(X, lowercase) when OTP version is always > 25
     string:lowercase(binary:encode_hex(X)).
+
+%% @doc default derived key length for PBKDF2, backword compatible with the old implementation in pbkdf2
+-spec dk_length(pbkdf2_mac_fun()) -> non_neg_integer().
+dk_length(md4) -> 16;
+dk_length(md5) -> 16;
+dk_length(ripemd160) -> 20;
```

Comment:
these three digests are not supported by cyrpto:pbkdf2_hmac

Reply:
bg: these are not FIPS_DIGEST, OTP disable them on purpose although openssl supports them.